### ISSUE TITLE
Unpredictable string to ip transformations results

### ISSUE THREAD

**Issue by vasilevalex** [OpenSIPS/opensips#1568]
Hi everybody.
OpenSIPS 2.4.3 this script:

```
$var(ip_1_s) = "[2001:6c8:5e00:1::249]";
$var(ip_1_b) = $(var(ip_1_s){ip.pton}); # Save binary IP to variable
$var(ip_1) = $(var(ip_1_s){ip.pton}{ip.ntop}); # Save string from binary IP to variable
$var(ip_2_s) = "[2001:6c8:5e00:1::250]";
$var(ip_2_b) = $(var(ip_2_s){ip.pton}); # Save binary IP to variable
$var(ip_2) = $(var(ip_2_s){ip.pton}{ip.ntop}); # Save string from binary IP to variable
xlog("IP1 is $(var(ip_1))");
xlog("IP2 is $(var(ip_2))");
if ($(var(ip_1_s){ip.pton}) == $(var(ip_2_s){ip.pton})) {
        xlog("Binary addresses are EQUAL");
} else {
        xlog("Binary addresses are DIFFERENT");
}
if ($(var(ip_1_s){ip.pton}{ip.ntop}) == $(var(ip_2_s){ip.pton}{ip.ntop})) {
        xlog("Binary to string addresses are EQUAL");
} else {
        xlog("Binary to string addresses are DIFFERENT");
}
if ($var(ip_1_b) == $var(ip_2_b)) {
        xlog("Variable saved Binary addresses are EQUAL");
} else {
        xlog("Variable saved Binary addresses are DIFFERENT");
}
if ($var(ip_1) == $var(ip_2)) {
        xlog("Variable saved Binary to string addresses are EQUAL");
} else {
        xlog("Variable saved Binary to string addresses are DIFFERENT");
}
```
All 4 ifs should return false, but:

```
Dec 13 11:23:09 test64 /usr/sbin/opensips[6063]: IP1 is 2001:6C8:5E00:1:0:0:0:249 af is INET6
Dec 13 11:23:09 test64 /usr/sbin/opensips[6063]: IP2 is 2001:6C8:5E00:1:0:0:0:250 af is INET6
Dec 13 11:23:09 test64 /usr/sbin/opensips[6063]: Binary addresses are EQUAL
Dec 13 11:23:09 test64 /usr/sbin/opensips[6063]: Binary to string addresses are EQUAL
Dec 13 11:23:09 test64 /usr/sbin/opensips[6063]: Variable saved Binary addresses are EQUAL
Dec 13 11:23:09 test64 /usr/sbin/opensips[6063]: Variable saved Binary to string addresses are DIFFERENT
```

**Comment by razvancrainea** [comment]
@vasilevalex I've looked into this and unfortunately the issue is not that easy - due to the way transformations work, this kind of issue can happen with other transformations too.
The main problem is that almost all transformations are using static buffers - this means that in an `if` you are basically not comparing the two values resulted by the transformation, but always the same buffer, that's why almost all `equal`, except the last one, where you only compare the values - properly stored in different buffers.

Long story short, you cannot use transformations as you do, you'll have to store each output in a different value. For example:
```
if ($(var(ip_1_s){ip.pton}) == $(var(ip_2_s){ip.pton}))
# should become
$var(ip_1_s_pton) = $(var(ip_1_s){ip.pton});
$var(ip_2_s_pton) = $(var(ip_2_s){ip.pton});
if ($var(ip_1_s_pton) == $var(ip_2_s_pton))
```

I have opened a new ticket #1572 that describes better the problem. I will now close this one.

**Comment by vasilevalex** [comment]
Thanks, @razvancrainea. Now I understand behavior and will use transformations according it.
